src/route/dataset-route.ts

```diff
@@ -0,0 +1,176 @@
+    }
+    const dataset = Dataset.createDataset(internalName, 'BetaUser');
+    const saved_dataset_record = await dataset.save();
+    const uploadDTO = await uploadCSV(req.file?.buffer, saved_dataset_record);
+    if (!uploadDTO.success) {
+        res.status(400);
+    }
+    res.json(uploadDTO);
+});
+
+apiRoute.get('/', async (req, res) => {
+    const datasets = await Dataset.find();
+    const fileList: FileDescription[] = [];
+    for (const dataset of datasets) {
+        fileList.push({
+            internal_name: dataset.internalName,
+            id: dataset.id
+        });
+    }
+    res.json({ filelist: fileList });
+});
+
+apiRoute.get('/:dataset', async (req, res) => {
+    const datasetID = req.params.dataset;
+    if (datasetID === undefined || datasetID === null) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... file is null or undefined' });
+        return;
+    }
+    const dataset = await Dataset.findOneBy({ id: datasetID });
+    if (!dataset) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... dataset id is invalid' });
+        return;
+    }
+    const datafiles = await dataset.datafiles;
+    if (datafiles.length < 1) {
+        res.status(404);
+        res.json({ message: 'Dataset has no datafiles attached' });
+        return;
+    }
+    const dto = await datasetToDatasetDTO(dataset);
+    res.json(dto);
+});
+
+apiRoute.get('/:dataset/csv', async (req, res) => {
+    const dataLakeService = new DataLakeService();
+    const datasetID = req.params.dataset;
+    if (datasetID === undefined || datasetID === null) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... You must specify a dataset ID' });
+        return;
+    }
+    const dataset = await Dataset.findOneBy({ id: datasetID });
+    if (dataset === undefined || dataset === null) {
+        res.status(404);
+        res.json({ message: 'Dataset not found... Dataset ID not found in Database' });
+        return;
+    }
+    const datafiles = await dataset.datafiles;
+    const fileToDownload: Datafile | undefined = datafiles
+        .sort(
+            (first: Datafile, second: Datafile) =>
+                new Date(second.creationDate).getTime() - new Date(first.creationDate).getTime()
+        )
+        .shift();
+    if (fileToDownload === undefined || fileToDownload === null) {
+        res.status(404);
+        res.json({ message: 'Dataset has no file attached' });
+        return;
+    }
+    const file = await dataLakeService.downloadFile(`${fileToDownload.id}.csv`);
+    if (file === undefined || file === null) {
+        res.status(404);
+        res.json({ message: 'File not found... file is null or undefined' });
+        return;
+    }
+    res.setHeader('Content-Length', file.length);
+    res.setHeader('Content-Type', 'text/csv');
+    res.setHeader('Content-Disposition', `attachment; filename=${fileToDownload.id}.csv`);
+    res.write(file, 'binary');
```

Comment:
Can we stream the file in blocks from the datalake to the response object so that we don't have to have the whole thing in memory at once (which could potentially be very large).

Reply:
Looking at the documentation for Datalake probably but again this will be subject to what specifically we want to return in the future.   Happy to iterate on this as time allows.